model/riscv_step.sail

```diff
@@ -15,36 +15,74 @@ union Step = {
-
-    hart_state = HART_ACTIVE();
-    // "When TW=1, then if WFI is executed in any
-    // less-privileged mode, and it does not complete within an
-    // implementation-specific, bounded time limit, the WFI
-    // instruction causes an illegal-instruction exception."
-    if   (cur_privilege == Machine | mstatus[TW] == 0b0)
-    then Step_Execute(Retire_Success(), instbits)
-    else Step_Execute(Illegal_Instruction(), instbits)
-  } else {
-    if   get_config_print_instr()
-    then print_instr("remaining in WAIT state at PC " ^ BitStr(PC));
-    Step_Waiting()
+function run_hart_waiting(step_no : int, wr: WaitReason, instbits : instbits, exit_wait : bool) -> Step = {
+  match (shouldWakeForInterrupt(), wr, valid_reservation(), exit_wait) {
+    // successfully interrupted waits
+    (true, _, _, _) => {
+      if  get_config_print_instr()
+      then print_instr("interrupt exit from " ^ wait_name(wr) ^ " state at PC " ^ BitStr(PC));
+
+      // The waiting instruction retires successfully.  The
+      // pending interrupts will be handled in the next step.
+      hart_state = HART_ACTIVE();
+      Step_Execute(Retire_Success(), instbits)
+    },
+    (_, WAIT_WRS_STO, false, _) => {
+      if   get_config_print_instr()
+      then print_instr("reservation invalid exit from " ^ wait_name(WAIT_WRS_STO) ^ " state at PC " ^ BitStr(PC));
+
+      hart_state = HART_ACTIVE();
+      Step_Execute(Retire_Success(), instbits)
+    },
+    (_, WAIT_WRS_NTO, false, _) => {
+      if   get_config_print_instr()
+      then print_instr("reservation invalid exit from " ^ wait_name(WAIT_WRS_NTO) ^ " state at PC " ^ BitStr(PC));
+
+      hart_state = HART_ACTIVE();
+      Step_Execute(Retire_Success(), instbits)
+    },
+    // Transition out of waiting states as instructed.
+    (_, WAIT_WFI, _, true) => {
+      if   get_config_print_instr()
+      then print_instr("forced exit from " ^ wait_name(WAIT_WFI) ^ " state at PC " ^ BitStr(PC));
+
+      hart_state = HART_ACTIVE();
+      // "When TW=1, then if WFI is executed in any
+      // less-privileged mode, and it does not complete within an
+      // implementation-specific, bounded time limit, the WFI
+      // instruction causes an illegal-instruction exception."
+      if   (cur_privilege == Machine | mstatus[TW] == 0b0)
+      then Step_Execute(Retire_Success(), instbits)
+      else Step_Execute(Illegal_Instruction(), instbits)
+    },
+    (_, WAIT_WRS_STO, _, true) => {
+      if   get_config_print_instr()
+      then print_instr("timed-out exit from " ^ wait_name(WAIT_WRS_STO) ^ " state at PC " ^ BitStr(PC));
+
+      hart_state = HART_ACTIVE();
+      Step_Execute(Retire_Success(), instbits)
+    },
+    (_, WAIT_WRS_NTO, _, true) => {
+      if   get_config_print_instr()
+      then print_instr("timed-out exit from " ^ wait_name(WAIT_WRS_NTO) ^ " state at PC " ^ BitStr(PC));
+
+      hart_state = HART_ACTIVE();
+      if   cur_privilege == Machine
```

Comment:
This is the same logic as line 57 but written in a more complex way. We should copy lines 57-59 here.

Reply:
Ok, I copied 57-59 for now, but as the TODO comment for the `else` case stated, it will have to be changed for VS/VU modes, and then will differ again.  WFI and WRS.NTO do behave differently with respect to causing `Illegal-Instruction`, as the spec explicitly notes.